Adds `getArgs :: IO [OsString]`

tbidne opened the pull request:
Resolves #338.

tomjaguarpaw commented:
This seems all fine to me, but I don't really "own" this library and don't know who does, so I don't know who gives final sign-off.

tbidne commented on System/Process/Env/OsString.hs:
```diff
@@ -0,0 +1,45 @@
+{-# OPTIONS_GHC -Wno-deprecations #-}
+#endif
+
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  System.Process.Env.OsString
+-- Copyright   :  (c) The University of Glasgow 2004-2008
+-- License     :  BSD-style (see the file libraries/base/LICENSE)
+--
+-- Maintainer  :  libraries@haskell.org
+-- Stability   :  experimental
+-- Portability :  non-portable (requires concurrency)
+--
+-- Miscellaneous information about the system environment.
+--
+-- @since X.Y.Z
+--
+-----------------------------------------------------------------------------
```
Indeed, the copied bit about portability (`non-portable (requires concurrency)`) also seems dubious if the comment is meant to apply to this module only. Happy to receive guidance here.

Bodigrim replied:
I'd skip all metadata fields here - they do not seem to add anything helpful.

tomjaguarpaw replied:
I second that.

Bodigrim commented:
@tomjaguarpaw I'm pretty sure that you have all necessary authority to make decisions here, since Michael passed the torch. Otherwise you can reach to CLC for advice, but this PR does not look particularly controversial to me.

tbidne commented:
Two questions I have:

1. Do we need to do anything about js or wasm? I see `javascript_HOST_ARCH` and `wasm32_HOST_ARCH` pragmas used in various places. I tried the legacy string `getArgs` and unix `getArgs` on the js backend, and both give empty args without an error, so maybe it's fine?

2. What of the rest of `System.Environment`? I don't mean to feature-creep, but I think it would be helpful to have an idea of what is to become of the rest of that module. Does the argument in favor of adding `getArgs` here ("args pertaining to the current process") also apply to the environment and/or program name/path? If so, everything in that module could eventually go here.

    Right now, `getEnv :: String -> IO String` and `getEnvironment :: IO [(String, String)]` exist in `unix` and `Win32` for `PosixString` and `WindowsString`, respectively, so these could also be easily shimmed here in this PR. The others do not exist yet.

tomjaguarpaw commented on System/Process/Env/OsString.hs:
```diff
@@ -0,0 +1,32 @@
+
+#if !MIN_VERSION_filepath(1,5,0)
+-- Silence the warning on System.OsString.Internal.Types import. It exists
+-- (deprecated) in filepath 1.4.100.0 && < 1.5.0.0. For filepath >= 1.5,
```
```suggestion
-- (deprecated) in filepath >= 1.4.100.0 && < 1.5.0.0. For filepath >= 1.5,
```

Is there a missing `>=` here?

tomjaguarpaw replied:
This doesn't seem to have been resolved. Has a push gone astray?

tbidne replied:
Possibly? I removed the entire comment per [your comment](https://github.com/haskell/process/pull/339#discussion_r2034063116). In any case, it's gone now.

tomjaguarpaw commented on System/Process/Env/OsString.hs:
```diff
@@ -0,0 +1,32 @@
+{-# LANGUAGE CPP #-}
+
+#if !MIN_VERSION_filepath(1,5,0)
+-- Silence the warning on System.OsString.Internal.Types import. It exists
+-- (deprecated) in filepath 1.4.100.0 && < 1.5.0.0. For filepath >= 1.5,
```
I don't see any deprecation notices here:

* https://hackage.haskell.org/package/filepath-1.4.100.0/docs/System-OsString-Internal-Types.html
* https://hackage.haskell.org/package/filepath-1.4.100.1/docs/System-OsString-Internal-Types.html
* https://hackage.haskell.org/package/filepath-1.4.100.2/docs/System-OsString-Internal-Types.html
* https://hackage.haskell.org/package/filepath-1.4.100.3/docs/System-OsString-Internal-Types.html
* https://hackage.haskell.org/package/filepath-1.4.100.4/docs/System-OsString-Internal-Types.html

and compiling with `cabal repl -b filepath==1.4.100.0` and this CPP removed does not emit a warning.

And in any case, I'm personally not fan of using CPP to suppress warnings. Do we really need this CPP?

tbidne replied:
Ah, it's in the 300+ series e.g. https://hackage.haskell.org/package/filepath-1.4.301.0/docs/System-OsString-Internal-Types.html.

In any case, no, it's not necessary. Removed.

tomjaguarpaw replied:
This doesn't seem to have been removed. Has a push gone astray?

tomjaguarpaw commented on process.cabal:
```diff
@@ -90,5 +97,10 @@ library
                    deepseq   >= 1.1 && < 1.6
+
+    if flag(os-string)
+        build-depends: filepath >= 1.5.0.0 && <1.6,
+                       os-string >= 2.0.0 && <2.1
```
I don't really get what's going on here. This seems bad:

```
cabal repl -b os-string==2.0.7
___________________________________________________________________________________________________________Sat 5 Apr 13:08:41 BST 2025
Configuration is affected by the following files:
- cabal.project
Warning: The package list for 'hackage.haskell.org' is 34 days old.
Run 'cabal update' to get the latest list of available packages.
Resolving dependencies...
Build profile: -w ghc-9.6.6 -O1
In order, the following will be built (use -v for more details):
 - process-1.6.25.0 (interactive) (first run)
Preprocessing library for process-1.6.25.0...
GHCi, version 9.6.6: https://www.haskell.org/ghc/  :? for help
[1 of 9] Compiling System.Process.Common ( System/Process/Common.hs, interpreted )
[2 of 9] Compiling System.Process.Env ( System/Process/Env.hs, interpreted )
[3 of 9] Compiling System.Process.Env.OsString ( System/Process/Env/OsString.hs, interpreted )

System/Process/Env/OsString.hs:23:1: error:
    Ambiguous module name ‘System.OsString.Internal.Types’:
      it was found in multiple packages:
      filepath-1.4.300.1 os-string-2.0.7
   |
23 | import System.OsString.Internal.Types
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
[4 of 9] Compiling System.Process.Posix ( System/Process/Posix.hs, interpreted )
[5 of 9] Compiling System.Process.Internals ( System/Process/Internals.hs, interpreted )
[6 of 9] Compiling System.Process.CommunicationHandle.Internal ( dist-newstyle/build/x86_64-linux/ghc-9.6.6/process-1.6.25.0/build/System/Process/CommunicationHandle/Internal.hs, interpreted )
[7 of 9] Compiling System.Process   ( System/Process.hs, interpreted )
[8 of 9] Compiling System.Process.CommunicationHandle ( System/Process/CommunicationHandle.hs, interpreted )
[9 of 9] Compiling System.Cmd       ( System/Cmd.hs, interpreted )
Failed, 8 modules loaded.
```

tomjaguarpaw replied:
This seems to fix things:

```diff
@ process.cabal:106 @ library
         build-depends: filepath >= 1.5.0.0 && <1.6,
                        os-string >= 2.0.0 && <2.1
     else
-        build-depends: filepath >= 1.4.100.0 && < 1.5.0.0
+        build-depends: filepath >= 1.4.100.0 && < 1.5.0.0,
+                       os-string < 0
```

```
cabal repl -b os-string==2.0.7
...
<succeeds>
```

That is to say, if you chose the "os-string: False" branch then you're not allowed to also use `os-string`-the-package.

tomjaguarpaw replied:
(I suppose another workaround would be package imports.)

tbidne replied:
`os-string < 0` seems like a good idea. @hasufell?

hasufell replied:
I'm actually not quite sure what the implication of this constraint is. But we should do it consistently for all boot packages.

E.g. there have been odd interactions with `--allow-newer`, which this is likely prone to as well.

`-b` in cabal is kind of a hack. You can break stuff arbitrarily, if you add any package via `-b` that exports the same modules as another package.

I actually think package imports might be a better option... e.g.

```hs
#if MIN_VERSION_filepath(1, 5, 0)
import "os-string" System.OsString.Internal.Types (OsString(OsString))
#else
import "filepath" System.OsString.Internal.Types (OsString(OsString))
#endif
```

hasufell replied:
@Bodigrim

Bodigrim replied:
I'd prefer `PackageImports`, it's unambiguous.

tbidne replied:
So do we keep the cabal flag too? We still need to ensure `os-string` is in `build-depends` when `filepath >= 1.5`.

I guess the alternative is to unconditionally require `os-string`. It will be superfluous in the `filepath < 1.5` case, but maybe that's better than mucking about with flags.

hasufell replied:
We keep the flag and the constraints the same way other packages do (as in: without the `< os-string-0`).

tomjaguarpaw replied:
Yeah, I think that's the right decision. `PackageImports` actually solves the particular problem, whereas `< 0` is a cute hack but only solves the problem accidentally.

tomjaguarpaw commented on System/Process/Env/OsString.hs:
```diff
@@ -0,0 +1,32 @@
+#if defined(mingw32_HOST_OS)
+import qualified System.Win32.WindowsString.Console as Platform
+import System.OsString.Internal.Types
+  ( OsString(OsString), WindowsString(WindowsString) )
+#else
+import System.OsString.Internal.Types
+  ( OsString(OsString), PosixString(PosixString) )
```
I don't understand why `WindowsString(WindowsString)` and `PosixString(PosixString)` are imported. Are they used? I don't see where. One might think they're needed for the `coerce` to work, but I think that only needs the `OsString` constructor. However, they don't emit redundant import warnings either, so I'm confused. In fact, if I change the import to

```.hs
import System.OsString.Internal.Types
 (OsString (OsString), PosixString)
```

then I *do* get a redundant import warning. Is GHC's redundant import checker just wrong when `coerce` is in play?

(I guess ideally we'd use [`coercionToPlatformTypes`](https://hackage.haskell.org/package/os-string-2.0.7/docs/System-OsString-Internal-Types.html#v:coercionToPlatformTypes). We shouldn't have to be diving into `os-string` internals here. But maybe we can't use it because we want to work with older `filepath`s?)

tbidne replied:
You are right. I reflexively imported `PosixString/WindowsString` as I thought it was needed for the coercion, but we only need `OsString`'s constructor.

The `OsString` constructor only exists in internal modules, so I'm not sure what else we can do. The coercion function you linked is interesting, but I think here cpp is preferable to runtime pattern matching.

tomjaguarpaw replied:
The coercion function probably gets inlined and eliminated (if it doesn't that's a bug and we do whatever is required to fix it), but it's not available in older versions of our dependencies, so we'll have to do without it here.

tomjaguarpaw commented:
> What of the rest of `System.Environment`?

I think if we're going to include one of its elements we may as well include them all.

tomjaguarpaw commented on System/Process/Env/OsString.hs:
```diff
@@ -0,0 +1,36 @@
+-- | 'System.Environment.getEnv' for 'OsString'.
+--
+-- @since X.Y.Z
+getEnv :: OsString -> IO (Maybe OsString)
+getEnv = coerce . Platform.getEnv . coerce
```
```suggestion
getEnv :: OsString -> IO (Maybe OsString)
getEnv = coerce Platform.getEnv
```

The point of `coerce` is that you just `coerce`!

tbidne commented:
> > What of the rest of System.Environment?
>
> I think if we're going to include one of its elements we may as well include them all.

Okay, I've added the remaining functions (`getEnv` and `getEnvironment`) that currently exist in all 3 (`System.Environment`, `Win32`, and `unix`). Here is a list of outstanding functions that should probably also be here, once they exist in `Win32` and `unix`:

- `Env.getProgName :: IO OsString`
- `Env.executablePath :: Maybe (IO (Maybe OsString))`
- `Env.getExecutablePath :: IO OsString`
- `Env.lookupEnv :: OsString -> IO (Maybe OsString)`
- `Env.setEnv :: OsString -> OsString -> IO ()`
- `Env.unsetEnv :: OsString -> IO ()`
- `Env.withArgs :: [OsString] -> IO a -> IO a`
- `Env.withProgName :: String -> IO a -> IO a`

tbidne commented:
Some bikeshedding:

1. `System.Process.Env` (current) or `System.Process.Environment` (like `System.Environment`)?
2. Generic haddocks e.g. `'System.Environment.getArgs' for 'OsString'` or copied from `System.Environment`?

tomjaguarpaw commented:
1. I'd prefer keeping it as similar as possible to existing module names, so unless there's also something called `Env` that we're copying, I'd stick to `Environment`

tbidne commented:
Okay, renamed `Env -> Environment` and added the `MIN_VERSION_filepath(1, 5, 0)` cpp. Two CI questions:

1. Do we care about the failing `GHC 8.4 + Windows` job, due to [this](https://github.com/haskell/win32/pull/229)? To recap, the problem is that we now require a recent version of `Win32`, which unfortunately added `NumericUnderscores`, causing `GHC < 8.6.1` to fail.

    Ideally that PR would be merged and released, so CI here would go back to green once the solver starts picking the new release. But it's been open for a year, so that might not be anytime soon. The combination could be excluded here, if we want.

2. Do we want to test the flag on CI in some capacity? It's easy enough to add a new matrix for `os-string/no os-string`, though of course that basically doubles the number of jobs, from 41 to 82 (it's not _quite_ that bad, since some older combinations should be excluded). This is well within the free tier limit (256 I think), but still, it will make CI take longer.

tomjaguarpaw commented on changelog.md:
```diff
@@ -1,5 +1,10 @@
 
+## X.Y.Z
+
+* Add `System.Process.Env` and `System.Process.Env.OsString`. Bumps
+  `filepath >= 1.4.100.0`.
```
This is now outdated. `s/Env/Environment/`?

tbidne replied:
Good catch, thanks.

tomjaguarpaw commented on process.cabal:
```diff
@@ -56,19 +61,21 @@ library
         System.Process
         System.Process.CommunicationHandle
         System.Process.CommunicationHandle.Internal
+        System.Process.Environment
+        System.Process.Environment.OsString
```
I don't understand why we have two modules which export the exact same thing. What's that about?

hasufell replied:
They don't. One uses `OsString`, one doesn't.

See:

* https://hackage.haskell.org/package/filepath-1.5.4.0/docs/System-FilePath.html
* https://hackage.haskell.org/package/filepath-1.5.4.0/docs/System-OsPath.html

And:

* https://hackage.haskell.org/package/directory-1.3.9.0/docs/System-Directory.html
* https://hackage.haskell.org/package/directory-1.3.9.0/docs/System-Directory-OsPath.html

tomjaguarpaw replied:
Oh, OK, one of them re-exports stuff from `System.Environment`. Why?

hasufell replied:
To have symmetric API. Otherwise you can only get those functions from process if you use OsString.

tomjaguarpaw replied:
I don't understand why one would want to get them from `process`. They're already in `base`.

hasufell replied:
There is no hard technical reason.

I think re-export makes sense, unless it's already in Prelude.

tbidne replied:
As hasufell mentioned, I did this to be consistent with the other libraries that offer `OsPath/OsString` versions of an existing `FilePath` API (see also [path](https://hackage.haskell.org/package/path)).

However, it's fair to say that the situation here is slightly different, since those libraries merely translated an API that already existed within the library, whereas here we are adding an API that is new for the library itself.

Personally, I am indifferent on including a `FilePath` version, as I don't plan to use it myself. Happy to go with consensus.

tomjaguarpaw replied:
It's not a key part of the PR and it's easier to add it later than it is to remove it, so let's be cautious and omit it for now.

tomjaguarpaw commented:
I suspect it's probably fine to give up on GHC 8.4.

tbidne committed:
Add System.Process.OsString module

Bumps lower bounds for base, filepath, unix, and Win32. Drops support
for GHC < 8.6, as required by newer Win32.

tomjaguarpaw commented:
Alright, this is looking good! @tbidne, is this ready to merge in your opinion?

Regarding testing more in the CI matrix, I'm happy for that to be added at a later date.

tbidne commented:
LGTM.

tomjaguarpaw commented:
Released as https://hackage.haskell.org/package/process-1.6.26.0

Thanks @tbidne and @hasufell
